src/InteractiveWindow/EditorTest/HistoryTests.cs

```diff
@@ -0,0 +1,308 @@
+        {
+            var buffer = _testHost.ExportProvider.GetExport<ITextBufferFactoryService>().Value.CreateTextBuffer();
+            buffer.Insert(0, "0123456789");
+            var snapshot = buffer.CurrentSnapshot;
+
+            var history = new History(maxLength: 0);
+
+            history.Add(new SnapshotSpan(snapshot, new Span(0, 1)));
+            Assert.Empty(GetHistoryEntries(history));
+
+            history = new History(maxLength: 1);
+            history.Add(new SnapshotSpan(snapshot, new Span(0, 1)));
+            Assert.Equal(new[] { "0" }, GetHistoryEntries(history));
+            history.Add(new SnapshotSpan(snapshot, new Span(1, 1)));
+            Assert.Equal(new[] { "1" }, GetHistoryEntries(history)); // Oldest entry is dropped.
+
+            history = new History(maxLength: 2);
+            history.Add(new SnapshotSpan(snapshot, new Span(0, 1)));
+            Assert.Equal(new[] { "0" }, GetHistoryEntries(history));
+            history.Add(new SnapshotSpan(snapshot, new Span(1, 1)));
+            Assert.Equal(new[] { "0", "1" }, GetHistoryEntries(history));
+            history.Add(new SnapshotSpan(snapshot, new Span(2, 1)));
+            Assert.Equal(new[] { "1", "2" }, GetHistoryEntries(history)); // Oldest entry is dropped.
+        }
+
+        [Fact]
+        public void TestLast()
+        {
+            for (int i = 0; i < 3; i++)
+            {
+                var text = i.ToString();
+                AddEntries(text);
+                Assert.Equal(text, _history.Last.Text);
+            }
+        }
+
+        [Fact]
+        public void TestForgetOriginalBuffers()
+        {
+            var entries = new[] { "1", "2", "3" };
+            AddEntries(entries);
+
+            Assert.Equal(entries, HistoryEntries);
+            AssertEx.All(_history.Items, e => e.OriginalSpan.HasValue);
+
+            _history.ForgetOriginalBuffers();
+
+            Assert.Equal(entries, HistoryEntries);
+            AssertEx.None(_history.Items, e => e.OriginalSpan.HasValue);
+        }
+
+        [Fact]
+        public void TestDuplicateEntries()
+        {
+            AddEntries("1", "1", "2", "2", "1", "2", "2", "1");
+            Assert.Equal(new[] { "1", "2", "1", "2", "1" }, HistoryEntries);
+        }
+
+        [Fact]
+        public void TestPrevious()
+        {
+            AddEntries("1", "2", "3");
+
+            Test(
+                new Step(() => _history.GetPrevious(null), "3"),
+                new Step(() => _history.GetPrevious(null), "2"),
+                new Step(() => _history.GetPrevious(null), "1"),
+                new Step(() => _history.GetPrevious(null), null),
+                new Step(() => _history.GetPrevious(null), null));
+        }
+
+        [Fact]
+        public void TestNext()
+        {
+            AddEntries("1", "2", "3");
+
+            Test(
+                new Step(() => _history.GetNext(null), null),
+                new Step(() => _history.GetNext(null), null),
```

Comment:
Could this be rewritten without Step?
Test(_history.GetPrevious(null), "3");
Test(_history.GetPrevious(null), "2");
...